algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/molecular/IsMmrDeficient.kt

```diff
@@ -1,26 +1,57 @@
 package com.hartwig.actin.algo.evaluation.molecular
 
 import com.hartwig.actin.algo.evaluation.EvaluationFactory
+import com.hartwig.actin.algo.evaluation.EvaluationFunction
+import com.hartwig.actin.algo.evaluation.IhcTestEvaluation
 import com.hartwig.actin.algo.evaluation.util.Format
+import com.hartwig.actin.datamodel.PatientRecord
 import com.hartwig.actin.datamodel.algo.Evaluation
-import com.hartwig.actin.datamodel.molecular.MolecularTest
+import com.hartwig.actin.datamodel.molecular.MolecularHistory
 import com.hartwig.actin.datamodel.molecular.driver.GeneAlteration
+import com.hartwig.actin.molecular.filter.MolecularTestFilter
 import com.hartwig.actin.molecular.util.GeneConstants
 import com.hartwig.actin.molecular.util.MolecularCharacteristicEvents
 import java.time.LocalDate
 
-class IsMicrosatelliteUnstable(maxTestAge: LocalDate? = null) : MolecularEvaluationFunction(maxTestAge) {
+class IsMmrDeficient(private val maxTestAge: LocalDate? = null) : EvaluationFunction {
+
+    override fun evaluate(record: PatientRecord): Evaluation {
+        val ihcTestEvaluation = IhcTestEvaluation.create("MMR", record.ihcTests)
+        val isMmrDeficientIhcResult = ihcTestEvaluation.hasCertainPositiveResultsForItem()
+        val isMmrProficientIhcResult = ihcTestEvaluation.hasCertainNegativeResultsForItem()
+
+        val molecularTestFilter = MolecularTestFilter(maxTestAge, false)
+        val molecularHistory = MolecularHistory(molecularTestFilter.apply(record.molecularTests))
+        val test = molecularHistory.latestOrangeMolecularRecord() ?: molecularHistory.allPanels()
+            .firstOrNull {
+                it.characteristics.microsatelliteStability != null || GeneConstants.MMR_GENES.any { gene ->
+                    it.testsGene(
+                        gene,
+                        any("")
+                    )
+                }
+            }
+
+        if (test == null && ihcTestEvaluation.filteredTests.isEmpty()) {
+            return EvaluationFactory.undetermined("No MMR deficiency test result", isMissingMolecularResultForEvaluation = true)
+        }
+
+        if (test == null) {
+            return when {
+                isMmrDeficientIhcResult -> EvaluationFactory.pass("Tumor is dMMR by IHC", inclusionEvents = setOf("MMR deficient"))
+                isMmrProficientIhcResult -> EvaluationFactory.fail("Tumor is not dMMR by IHC")
+                else -> EvaluationFactory.undetermined("Undetermined dMMR result by IHC")
+            }
+        }
 
-    override fun evaluate(test: MolecularTest): Evaluation {
         val drivers = test.drivers
-        val msiVariants = drivers.variants
-            .filter { variant -> variant.gene in GeneConstants.MMR_GENES && variant.isReportable }
+        val msiVariants = drivers.variants.filter { variant -> variant.gene in GeneConstants.MMR_GENES && variant.isReportable }
```

Comment:
(sorry this comment ended up in the wrong spot) if you groupBy the isBiallelic status here, you can avoid having to filter 3 times below for the different groups.